docs/Using-SRIOV-and-Clear-Containers.md

```diff
@@ -0,0 +1,237 @@
+
+```
+$ find /sys/kernel/iommu_groups/ -type l
+```
+
+For more details, see http://vfio.blogspot.com/2014/08/iommu-groups-inside-and-out.html
+
+### Updating the Host Kernel
+
+These directions are provided as an example, and are based on Ubuntu 16.04. The
+user will need to do something equivalent for their distribution.
+
+1.  Grab kernel sources
+
+```
+$ sudo apt-get install linux-source-4.10.0
+$ sudo apt-get install linux-headers-4.10.0
+$ cd /usr/src/linux-source-4.10.0/
+$ sudo tar -xvf linux-source-4.10.0.tar.bz2
+$ cd linux-source-4.10.0
+$ sudo apt-get install libssl-dev
+```
+
+2. Check the config and update if necessary
+
+```
+$ sudo cp /boot/config-4.8.0-36-generic .config
+$ # verify resulting .config does not have NOIOMMU set; ie: `CONFIG_VFIO_NOIOMMU` is not set
+$ # verify `CONFIG_VFIO_IOMMU_TYPE1`, `CONFIG_VFIO=m` and `CONFIG_VFIO_PCI=m` are set as well.
+$ sudo make olddefconfig 
+```
+
+You may choose to modify the kernel `Makefile` to add a unique indentifier 
+to the `EXTRAVERSION` variable proir to running make, so that when the
+kernel is installed and running, the command `uname -r` will indicate that the
+kernel being run has been customized.
+If your system is configured for automatic distribution updates, make sure
+you are running your customized kernel and not the new kernel
+installed with `apt-get update` or `dnf update`.
+
+3.  Build and install the kernel:
+
+```
+$ sudo make -j #and go get coffee
+$ sudo make modules -j 3
+$ sudo make modules_install
+$ sudo make install
+```
+
+4.  Edit grub to enable intel-iommu:
+
+```
+edit /etc/default/grub and add intel_iommu=on to cmdline:
+- GRUB_CMDLINE_LINUX=""
++ GRUB_CMDLINE_LINUX="intel_iommu=on"
+sudo update-grub
+```
+
+5.  Reboot and verify
+Host system should be ready now -- reboot and verify that expected cmdline and
+kernel version is booted (look at `/proc/cmdline` and `/proc/version`)
+
+```
+sudo reboot
+```
+
+6. Add `vfio-pci` module
+```
+sudo modprobe vfio-pci
+```
+
+7.  If necessary, add PCI quirk for SR-IOV NIC
+
+Verify your NIC appears in its own IOMMU group and no ther devices 
+appear in the same group, using the command below:
+```
+$ find /sys/kernel/iommu_groups/ -type l
+```
+
+In the rare case where your PCI NIC does not appear in its own group, 
+the NIC may not support ACS or you may have chosen to build and run an old kernel.
+Depending on your NIC and if it indeed enforces isolation, you may resolve this
+by adding a `pcie_acs_override=` option to your kernel command line and reboot.
+Detailed information about this option can be found at [PCIE-ACS-override-option](https://lkml.org/lkml/2013/5/30/513).
+
+## Setting up SR-IOV Device
+
+All of the prior sections are needed once to prepare the SR-IOV host systems.
+The following will be needed per boot in order to facilitate setting up a
+physical device's virtual functions.
+
+For SR-IOV, a physical device can create up to `sriov_totalvfs` virtual 
```

Comment:
Let's put the explanation of `sriov_totalvfs` before this sentence as currently: create up to `sriov_totalvfs` doesn't make sense